cmd/scaffold.go

```diff
@@ -952,6 +962,63 @@ func (fnb *FlowNodeBuilder) handleComponent(v namedComponentFunc, parentReady <-
 	return nil
 }
 
+// handleRestartableComponent constructs a component using the provided ReadyDoneFactory, and
+// registers a worker with the ComponentManager to be run when the node is started.
+//
+// Restartable Components are components that can be restarted after successfully handling
+// an irrecoverable error.
+//
+// Any irrecoverable errors thrown by the component will be passed to the provided error handler.
+func (fnb *FlowNodeBuilder) handleRestartableComponent(v namedComponentFunc, parentReady <-chan struct{}, started func()) error {
+	fnb.componentBuilder.AddWorker(func(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
+		// wait for the previous component to be ready before starting
+		if err := util.WaitClosed(ctx, parentReady); err != nil {
+			return
+		}
+
+		// Note: we're marking the worker routine ready before we even attempt to start the
+		// component. the idea behind a restartable component is that the node should not depend
+		// on it for safe operation, so the node does not need to wait for it to be ready.
+		ready()
+
+		// do not block serial startup. started can only be called once, so it cannot be called
+		// from within the componentFactory
+		started()
+
+		log := fnb.Logger.With().Str("component", v.name).Logger()
+
+		// This may be called multiple times if the component is restarted
+		componentFactory := func() (component.Component, error) {
+			c, err := v.fn(fnb.NodeConfig)
+			if err != nil {
+				return nil, err
+			}
+			log.Info().Msg("component initialization complete")
+
+			go func() {
+				if err := util.WaitClosed(ctx, c.Ready()); err != nil {
+					log.Info().Msg("component startup aborted")
+				} else {
+					log.Info().Msg("component startup complete")
+				}
+
+				<-ctx.Done()
+				log.Info().Msg("component shutdown started")
+			}()
+			return c.(component.Component), nil
+		}
+
+		err := component.RunComponent(ctx, componentFactory, v.errorHandler)
+		if err != nil && !errors.Is(err, ctx.Err()) {
+			ctx.Throw(fmt.Errorf("component %s shutdown unexpectedly: %w", v.name, err))
+		}
+
+		log.Info().Msg("component shutdown complete")
+	})
+
+	return nil
+}
+
```

Comment:
what component we use this for?

Reply:
Originally, this was used with the requester, but I removed it when I dropped the halting behavior. Since the code and tests were already written, I figured i'd just leave it in for the next time we want this functionality